src/System.IO.Pipelines/PipeCompletion.cs

```diff
@@ -1,30 +1,75 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System.Threading;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.IO.Pipelines
 {
     internal struct PipeCompletion
     {
+        private const int InitialCallbacksSize = 4;
         private static readonly Exception _completedNoException = new Exception();
 
-        private Exception _exception;
 #if COMPLETION_LOCATION_TRACKING
         private string _completionLocation;
 #endif
+        private Exception _exception;
+        private PipeCompletionCallback[] _callbacks;
+        private int _callbackCount;
+
+        public string Location
+        {
+            get
+            {
+#if COMPLETION_LOCATION_TRACKING
```

Comment:
Having just used this, I think we should have a single define for all "LOCATION_TRACKING".